src/librbd/internal.cc

```diff
@@ -1635,6 +1636,228 @@ int mirror_image_disable_internal(ImageCtx *ictx, bool force,
     return r;
   }
 
+  int trash_move(librados::IoCtx &io_ctx, rbd_trash_move_source_t source,
+                 const std::string &image_name, uint64_t delay) {
+    CephContext *cct((CephContext *)io_ctx.cct());
+    ldout(cct, 20) << "trash_move " << &io_ctx << " " << image_name
+                   << dendl;
+
+    std::string image_id;
+    ImageCtx *ictx = new ImageCtx(image_name, "", nullptr, io_ctx, false);
+    int r = ictx->state->open(true);
+    if (r < 0) {
+      ldout(cct, 2) << "error opening image: " << cpp_strerror(-r) << dendl;
+      delete ictx;
+      if (r != -ENOENT) {
+	return r;
+      }
+    } else {
+      if (ictx->old_format) {
+        ictx->state->close();
+        return -EOPNOTSUPP;
+      }
+
+      image_id = ictx->id;
+      ictx->owner_lock.get_read();
+      if (ictx->exclusive_lock != nullptr) {
+        r = ictx->operations->prepare_image_update();
+        if (r < 0 || !ictx->exclusive_lock->is_lock_owner()) {
+	  lderr(cct) << "cannot obtain exclusive lock - not removing" << dendl;
+	  ictx->owner_lock.put_read();
+	  ictx->state->close();
+          return -EBUSY;
+        }
+      }
+    }
+
+    BOOST_SCOPE_EXIT_ALL(ictx, cct) {
+      bool is_locked = ictx->exclusive_lock != nullptr &&
+                       ictx->exclusive_lock->is_lock_owner();
+      ictx->owner_lock.put_read();
+      if (is_locked) {
+        C_SaferCond ctx;
+        ictx->exclusive_lock->shut_down(&ctx);
```

Comment:
I had to release it because otherwise if I call shut_down while holding a read/write-lock over `owner_lock` it creates a deadlock between L1680 and https://github.com/ceph/ceph/blob/master/src/librbd/ExclusiveLock.cc#L281